pkg/controller/machine/machine_controller_phases.go

```diff
@@ -0,0 +1,269 @@
+		return errors.Errorf(
+			"Expected at least one of `Bootstrap.ConfigRef` or `Bootstrap.Data` to be populated for Machine %q in namespace %q",
+			m.Name, m.Namespace,
+		)
+	}
+
+	// If the Machine is being deleted and the ConfigRef is nil,
+	// set BootstrapReady to false and return.
+	if !m.DeletionTimestamp.IsZero() && m.Spec.Bootstrap.ConfigRef == nil {
+		m.Status.BootstrapReady = pointer.BoolPtr(false)
+		return nil
+	}
+
+	// If the Machine is not being deleted and Bootstrap's Data
+	// field is populate, return early.
+	if m.DeletionTimestamp.IsZero() && m.Spec.Bootstrap.Data != nil {
+		m.Status.BootstrapReady = pointer.BoolPtr(true)
+		return nil
+	}
+
+	// Call generic external reconciler.
+	bootstrapConfig, err := r.reconcileExternal(ctx, m, m.Spec.Bootstrap.ConfigRef)
+	if bootstrapConfig == nil && err == nil {
+		m.Status.BootstrapReady = pointer.BoolPtr(false)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// If the bootstrap config is being deleted, return early.
+	if !bootstrapConfig.GetDeletionTimestamp().IsZero() {
+		return nil
+	}
+
+	// Determine if the bootstrap provider is ready.
+	ready, err := r.isExternalReady(bootstrapConfig)
+	if err != nil {
+		return err
+	} else if !ready {
+		klog.V(3).Infof("Bootstrap provider for Machine %q in namespace %q is not ready, requeuing", m.Name, m.Namespace)
+		return &capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second}
+	}
+
+	// Get and set data from the bootstrap provider.
+	data, _, err := unstructured.NestedString(bootstrapConfig.Object, "status", "bootstrapData")
+	if err != nil {
+		return errors.Wrapf(err, "failed to retrieve data from bootstrap provider for Machine %q in namespace %q", m.Name, m.Namespace)
+	} else if data == "" {
+		return errors.Errorf("retrieved empty data from bootstrap provider for Machine %q in namespace %q", m.Name, m.Namespace)
+	}
+
+	m.Spec.Bootstrap.Data = pointer.StringPtr(data)
+	m.Status.BootstrapReady = pointer.BoolPtr(true)
+	return nil
+}
+
+// reconcileInfrastructure reconciles the Spec.InfrastructureRef object on a Machine.
+func (r *ReconcileMachine) reconcileInfrastructure(ctx context.Context, m *v1alpha2.Machine) error {
+	// Call generic external reconciler.
+	infraConfig, err := r.reconcileExternal(ctx, m, &m.Spec.InfrastructureRef)
+	if infraConfig == nil && err == nil {
+		m.Status.InfrastructureReady = pointer.BoolPtr(false)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	if (m.Status.InfrastructureReady != nil && *m.Status.InfrastructureReady) ||
+		!infraConfig.GetDeletionTimestamp().IsZero() {
+		return nil
+	}
+
+	// Determine if the infrastructure provider is ready
+	ready, err := r.isExternalReady(infraConfig)
+	if err != nil {
+		return err
+	} else if !ready {
+		klog.V(3).Infof("Infrastructure provider for Machine %q in namespace %q is not ready, requeuing", m.Name, m.Namespace)
+		return &capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second}
+	}
+
+	// Get addresses.
```

Comment:
Do we need this or can we use `ready` from above?

Reply:
This is following the proposal spec, if we want to change it, we should probably PR the proposal or open an issue

Reply:
I'd like us to consider changing what's in the proposal. I would prefer to have a single definition of infrastructure readiness, and I believe the value returned by `r.IsExternalReady` is sufficient.

Reply:
Yeah agreed, I'll remove it from here, although Infrastructure providers still need to specify addresses, am I correct?

Reply:
Only if we need addresses in CAPI?

Reply:
I don't know if CAPI needs it, but I think all infrastructure providers should give users addresses in a way or another. That's how the field ended up in the proposal